Tests to check correctness of WHERE clauses for PostgreSQL

espinafre opened the pull request:
When I submitted the improvement for handling PostgreSQL bigint/bigserial primary keys, preventing QGIS from casting those to text (PR #35162) I wrote some functional testcases in Python. Those tests only assured that my changes didn't break anything, but they didn't prove that the changes really worked as intended. I've checked that by looking at debug logs and the database logs. I didn't know then a way to formally check the generated SQL queries from inside the testing framework.

Now, while reviewing #36266, I saw that @benoitdm-oslandia came up with the idea of calling the QgsPostgresUtils::whereClause() method directly, while using an instance of QgsPostgresSharedData to pass the fields to the provider.

So, I stole his idea and wrote these tests, and I've gained some more insights about the QGIS provider in the process. For instance, I've found that in the case of compound keys, the original problem of QGIS casting integers to text persist; that's not a common use case, but still, I think it is worth a shot (thats what the FIXME comment on the last test is for).

Also, with these tests we cases we may rest more assured that there won't be regressions in this case.

espinafre committed:
Tests to check correctness of WHERE clauses for PostgreSQL

espinafre committed:
Changed tests after rebasing on top of 7120e5da328

That change made a few improvements, and uncovered a bug in handling
tables with composite primary keys if one component is an integer.

espinafre commented:
Hi, @nyalldawson , what do you think of this? @elpaso , can you comment, since you've worked o this recently? Travis, at least, likes it.

nyalldawson commented on tests/src/providers/testqgspostgresprovider.cpp:
```diff
@@ -190,5 +194,112 @@ void TestQgsPostgresProvider::testDecodeDateTimes()
+  f0.setType( QVariant::Int );
+  f0.setTypeName( "int" );
+
+  fields.clear();
+  pkAttrs.clear();
+  vlst.clear();
```
```suggestion
```

Not needed

espinafre replied:
Resolved by commit 07358a0 .

nyalldawson commented on tests/src/providers/testqgspostgresprovider.cpp:
```diff
@@ -190,5 +194,112 @@ void TestQgsPostgresProvider::testDecodeDateTimes()
+  pkAttrs.append( 0 );
+  vlst.append( 42 );
+
+  sdata = new QgsPostgresSharedData;
```
```suggestion
  std::shared_ptr< QgsPostgreSharedData > sdata( new QgsPostgresSharedData() );
```

espinafre replied:
Resolved by commit 07358a0 . Thanks for the hint!

nyalldawson reviewed (commented):
More tests are always appreciated! :smile:

espinafre committed:
Code review.

Cleaner usage of std::shared_ptr< QgsPostgresSharedData >; removed
useless calls to .clear() on first invocation of methods. Code reviewd
by Nyall Dawson.

elpaso commented on tests/src/providers/testqgspostgresprovider.cpp:
```diff
@@ -190,5 +194,108 @@ void TestQgsPostgresProvider::testDecodeDateTimes()
+
+  fields.append( f1 );
+  pkAttrs.append( 0 );
+  vlst.append( -9223372036854775800LL ); // way outside int4 range
```
did you mean way outside int8 ?

espinafre replied:
Outside int4, really, the value doesn't fit a 4-byte (32 bit) integer.

elpaso commented on tests/src/providers/testqgspostgresprovider.cpp:
```diff
@@ -190,5 +194,108 @@ void TestQgsPostgresProvider::testDecodeDateTimes()
+
+  fields.append( f2 );
+  pkAttrs.append( 0 );
+  vlst.append( 3.141592741 );
```
I like this one :+1:

espinafre replied:
A little homage to the most well rounded release of QGIS yet 😁

elpaso reviewed (approved):
LGTM Thanks!

espinafre commented:
Are we good to merge? Thanks!
